src/main/java/knightminer/simplytea/item/TeapotItem.java

```diff
@@ -4,69 +4,132 @@
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.ClipContext;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.AbstractCauldronBlock;
+import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.BucketPickup;
+import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluid;
+import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult.Type;
+import net.minecraftforge.common.Tags;
+import net.minecraftforge.common.capabilities.ForgeCapabilities;
+import net.minecraftforge.fluids.FluidActionResult;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidType;
+import net.minecraftforge.fluids.FluidUtil;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+
+import java.util.Optional;
 
 public class TeapotItem extends TooltipItem {
+
 	public TeapotItem(Properties props) {
 		super(props);
 	}
 
+	// using onItemUseFirst() instead of use() because many fluid tanks will consume the interaction before it gets to use()
 	@Override
-	public InteractionResultHolder<ItemStack> use(Level world, Player player, InteractionHand hand) {
-		ItemStack stack = player.getItemInHand(hand);
+	public InteractionResult onItemUseFirst(ItemStack stack, UseOnContext context)
+    {
+		Level world = context.getLevel();
+		if (world.isClientSide) {
+        	return InteractionResult.PASS;
+		}
+
+		Player player = context.getPlayer();
+		// Even though the context already has a blockhitresult, it does not include fluid blocks
 		BlockHitResult rayTrace = getPlayerPOVHitResult(world, player, ClipContext.Fluid.SOURCE_ONLY);
 		if (rayTrace.getType() == Type.BLOCK) {
 			BlockPos pos = rayTrace.getBlockPos();
 			BlockState state = world.getBlockState(pos);
+			Block block = state.getBlock();
+			Direction side = rayTrace.getDirection();
+
+			// unable to modify the block
+			if (!world.mayInteract(player, pos) || !player.mayUseItemAt(pos.relative(side), side, stack)) {
+				return InteractionResult.PASS;
+			}
+
+			// cauldron disabled in config
+			if (block instanceof AbstractCauldronBlock && !Config.SERVER.teapot.fillFromCauldron()) {
+				return InteractionResult.PASS;
+			}
+
+			Optional<Item> filledItem = Optional.empty();
+			Optional<SoundEvent> pickupSound = Optional.empty();
+			FluidState fluidState = state.getFluidState();
+
+			if (fluidState.isSource()) {
+				Fluid fluid = fluidState.getType();
+				// comparing with WATER fluid instead of WATER fluid tag, because many/most modded fluids are tagged as WATER
+				// to get water-like interactions even if the fluid is not water
+				if (fluid.isSame(Fluids.WATER)) {
```

Comment:
We should also support milk if Forge's `ForgeMod#MILK.isPresent()` or whatever the check is to see that the milk fluid is enabled.

Reply:
Between using the fluid handler on the teapot and using FluidUtil to pick up fluids, milk should be working without needing to specially check whether a milk fluid is enabled.